File: typescript/src/adapters/memory.ts
// pattern: Imperative Shell

import { GrpcqError } from "../error.js";
import type { Message } from "../message.js";
import type {
  ConsumeResult,
  MessageItem,
  QueueAdapter,
  Receipt,
} from "../types.js";

const enum ReceiptState {
  Pending = 0,
  Acked = 1,
  Nacked = 2,
}

export class MemoryAdapter implements QueueAdapter {
  private readonly queues = new Map<string, Message[]>();
  private readonly bufferSize: number;

  constructor(bufferSize = 1000) {
    this.bufferSize = bufferSize > 0 ? bufferSize : 1000;
  }

  async publish(queueName: string, messages: Message[]): Promise<void> {
    if (messages.length === 0) return;

    const queue = this.getOrCreateQueue(queueName);
    if (queue.length + messages.length > this.bufferSize) {
      throw GrpcqError.queueFull(queueName);
    }

    queue.push(...messages.map((m) => structuredClone(m)));
  }

  async consume(queueName: string, maxBatch: number): Promise<ConsumeResult> {
    const queue = this.getOrCreateQueue(queueName);
    const limit = Math.max(maxBatch, 1);
    const drained = queue.splice(0, limit);

    const items: MessageItem[] = drained.map((message) => ({
      message,
      receipt: this.createReceipt(queueName, message),
    }));

    return { items };
  }

  queueDepth(queueName: string): number {
    return this.queues.get(queueName)?.length ?? 0;
  }

  clear(): void {
    for (const queue of this.queues.values()) {
      queue.length = 0;
    }
  }

  private getOrCreateQueue(queueName: string): Message[] {
    let queue = this.queues.get(queueName);
    if (!queue) {
      queue = [];
      this.queues.set(queueName, queue);
    }
    return queue;
  }

  private createReceipt(queueName: string, message: Message): Receipt {
    let state = ReceiptState.Pending;

    return {
      ack: async () => {
        if (state === ReceiptState.Acked)
          throw GrpcqError.alreadyAcknowledged();
        if (state === ReceiptState.Nacked) throw GrpcqError.alreadyNacked();
        state = ReceiptState.Acked;
      },
      nack: async () => {
        if (state === ReceiptState.Acked)
          throw GrpcqError.alreadyAcknowledged();
        if (state === ReceiptState.Nacked) throw GrpcqError.alreadyNacked();
        state = ReceiptState.Nacked;

        const queue = this.getOrCreateQueue(queueName);
        if (queue.length < this.bufferSize) {
          queue.push(structuredClone(message));
        }
      },
    };
  }
}

File: rust/src/adapters/memory.rs
use std::{
    collections::{HashMap, VecDeque},
    sync::{Arc, Mutex},
};

use futures::{future::BoxFuture, FutureExt};

use crate::{
    core::{ConsumeResult, MessageItem, QueueAdapter, Receipt},
    Error, Message, Result, SharedReceipt,
};

#[derive(Clone)]
pub struct Adapter {
    inner: Arc<Inner>,
}

struct Inner {
    queues: Mutex<HashMap<String, VecDeque<Message>>>,
    buffer_size: usize,
}

impl Adapter {
    pub fn new(buffer_size: usize) -> Self {
        Self {
            inner: Arc::new(Inner {
                queues: Mutex::new(HashMap::new()),
                buffer_size: if buffer_size == 0 { 1000 } else { buffer_size },
            }),
        }
    }

    pub fn queue_depth(&self, queue_name: &str) -> usize {
        let queues = self
            .inner
            .queues
            .lock()
            .expect("memory adapter lock poisoned");
        queues.get(queue_name).map_or(0, VecDeque::len)
    }

    pub fn clear(&self) {
        let mut queues = self
            .inner
            .queues
            .lock()
            .expect("memory adapter lock poisoned");
        for queue in queues.values_mut() {
            queue.clear();
        }
    }
}

impl QueueAdapter for Adapter {
    fn publish<'a>(
        &'a self,
        queue_name: &'a str,
        messages: &'a [Message],
    ) -> BoxFuture<'a, Result<()>> {
        async move {
            if messages.is_empty() {
                return Ok(());
            }

            let mut queues = self
                .inner
                .queues
                .lock()
                .expect("memory adapter lock poisoned");
            let queue = queues.entry(queue_name.to_string()).or_default();

            if queue.len() + messages.len() > self.inner.buffer_size {
                return Err(Error::QueueFull {
                    queue_name: queue_name.to_string(),
                });
            }

            queue.extend(messages.iter().cloned());
            Ok(())
        }
        .boxed()
    }

    fn consume<'a>(
        &'a self,
        queue_name: &'a str,
        max_batch: usize,
    ) -> BoxFuture<'a, Result<ConsumeResult>> {
        async move {
            let drained = {
                let mut queues = self
                    .inner
                    .queues
                    .lock()
                    .expect("memory adapter lock poisoned");
                let queue = queues.entry(queue_name.to_string()).or_default();
                let limit = max_batch.max(1);
                let mut drained = Vec::with_capacity(limit);

                for _ in 0..limit {
                    match queue.pop_front() {
                        Some(message) => drained.push(message),
                        None => break,
                    }
                }

                drained
            };

            let items = drained
                .into_iter()
                .map(|message| MessageItem {
                    receipt: memory_receipt(self.inner.clone(), queue_name, message.clone()),
                    message,
                })
                .collect();

            Ok(ConsumeResult { items })
        }
        .boxed()
    }
}

fn memory_receipt(inner: Arc<Inner>, queue_name: &str, message: Message) -> SharedReceipt {
    Arc::new(MemoryReceipt {
        state: Mutex::new(ReceiptState::Pending),
        inner,
        queue_name: queue_name.to_string(),
        message,
    })
}

struct MemoryReceipt {
    state: Mutex<ReceiptState>,
    inner: Arc<Inner>,
    queue_name: String,
    message: Message,
}

enum ReceiptState {
    Pending,
    Acked,
    Nacked,
}

impl Receipt for MemoryReceipt {
    fn ack(&self) -> BoxFuture<'_, Result<()>> {
        async move {
            let mut state = self.state.lock().expect("memory receipt lock poisoned");
            match *state {
                ReceiptState::Pending => {
                    *state = ReceiptState::Acked;
                    Ok(())
                }
                ReceiptState::Acked => Err(Error::AlreadyAcknowledged),
                ReceiptState::Nacked => Err(Error::AlreadyNacked),
            }
        }
        .boxed()
    }

    fn nack(&self) -> BoxFuture<'_, Result<()>> {
        async move {
            {
                let mut state = self.state.lock().expect("memory receipt lock poisoned");
                match *state {
                    ReceiptState::Pending => {
                        *state = ReceiptState::Nacked;
                    }
                    ReceiptState::Acked => return Err(Error::AlreadyAcknowledged),
                    ReceiptState::Nacked => return Err(Error::AlreadyNacked),
                }
            }

            let mut queues = self
                .inner
                .queues
                .lock()
                .expect("memory adapter lock poisoned");
            let queue = queues.entry(self.queue_name.clone()).or_default();
            if queue.len() < self.inner.buffer_size {
                queue.push_back(self.message.clone());
            }

            Ok(())
        }
        .boxed()
    }
}
